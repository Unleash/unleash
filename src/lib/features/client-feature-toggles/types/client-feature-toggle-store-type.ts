import {
    IFeatureToggleClient,
    IFeatureToggleQuery,
} from '../../../types/model';
import { IGetAdminFeatures } from '../client-feature-toggle-store';

export interface IFeatureToggleClientStore {
    getClient(
        featureQuery: Partial<IFeatureToggleQuery>,
    ): Promise<IFeatureToggleClient[]>;

    getPlayground(
        featureQuery: Partial<IFeatureToggleQuery>,
    ): Promise<IFeatureToggleClient[]>;

    // @Deprecated
    getAdmin(params: IGetAdminFeatures): Promise<IFeatureToggleClient[]>;
}
