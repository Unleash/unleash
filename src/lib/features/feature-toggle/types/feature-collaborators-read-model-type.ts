export type Collaborator = {
    id: number;
    name: string;
    imageUrl: string;
};

export interface IFeatureCollaboratorsReadModel {
    getFeatureCollaborators(feature: string): Promise<Array<Collaborator>>;
}
