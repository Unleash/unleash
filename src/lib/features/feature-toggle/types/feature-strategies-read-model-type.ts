export interface IFeatureStrategiesReadModel {
    getMaxFeatureEnvironmentStrategies(): Promise<{
        feature: string;
        environment: string;
        count: number;
    } | null>;
    getMaxFeatureStrategies(): Promise<{
        feature: string;
        count: number;
    } | null>;
    getMaxConstraintValues(): Promise<{
        feature: string;
        environment: string;
        count: number;
    } | null>;
    getMaxConstraintsPerStrategy(): Promise<{
        feature: string;
        environment: string;
        count: number;
    } | null>;
}
