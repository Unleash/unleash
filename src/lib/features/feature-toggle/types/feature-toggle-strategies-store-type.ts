import {
    FeatureToggleWithEnvironment,
    IDependency,
    IFeatureOverview,
    IFeatureStrategy,
    IStrategyConfig,
    IVariant,
} from '../../../types/model';
import { Store } from '../../../types/stores/store';
import { IFeatureProjectUserParams } from '../feature-toggle-controller';

export interface FeatureConfigurationClient {
    name: string;
    type: string;
    enabled: boolean;
    project: string;
    stale: boolean;
    strategies: IStrategyConfig[];
    variants: IVariant[];
    dependencies?: IDependency[];
}

export interface IFeatureSearchParams {
    userId: number;
    searchParams?: string[];
    project?: string;
    segment?: string;
    createdAt?: string;
    state?: string;
    type?: string[];
    tag?: string;
    status?: string[][];
    offset: number;
    favoritesFirst?: boolean;
    limit: number;
    sortBy: string;
    sortOrder: 'asc' | 'desc';
}

export type IQueryOperator =
    | 'IS'
    | 'IS_NOT'
    | 'IS_ANY_OF'
    | 'IS_NONE_OF'
    | 'INCLUDE'
    | 'DO_NOT_INCLUDE'
    | 'INCLUDE_ALL_OF'
    | 'INCLUDE_ANY_OF'
    | 'EXCLUDE_IF_ANY_OF'
    | 'EXCLUDE_ALL'
    | 'IS_BEFORE'
    | 'IS_ON_OR_AFTER';

export interface IQueryParam {
    field: string;
    operator: IQueryOperator;
    values: string[];
}

export interface IFeatureStrategiesStore
    extends Store<IFeatureStrategy, string> {
    createStrategyFeatureEnv(
        strategyConfig: Omit<IFeatureStrategy, 'id' | 'createdAt'>,
    ): Promise<IFeatureStrategy>;

    removeAllStrategiesForFeatureEnv(
        featureName: string,
        environment: string,
    ): Promise<void>;

    getStrategiesForFeatureEnv(
        projectId: string,
        featureName: string,
        environment: string,
    ): Promise<IFeatureStrategy[]>;

    getFeatureToggleWithEnvs(
        featureName: string,
        userId?: number,
        archived?: boolean,
    ): Promise<FeatureToggleWithEnvironment>;

    getFeatureToggleWithVariantEnvs(
        featureName: string,
        userId?: number,
        archived?,
    ): Promise<FeatureToggleWithEnvironment>;

    getFeatureOverview(
        params: IFeatureProjectUserParams,
    ): Promise<IFeatureOverview[]>;

    getStrategyById(id: string): Promise<IFeatureStrategy>;

    updateStrategy(
        id: string,
        updates: Partial<IFeatureStrategy>,
    ): Promise<IFeatureStrategy>;

    deleteConfigurationsForProjectAndEnvironment(
        projectId: String,
        environment: String,
    ): Promise<void>;

    setProjectForStrategiesBelongingToFeature(
        featureName: string,
        newProjectId: string,
    ): Promise<void>;

    getStrategiesBySegment(segmentId: number): Promise<IFeatureStrategy[]>;

    getStrategiesByContextField(
        contextFieldName: string,
    ): Promise<IFeatureStrategy[]>;

    updateSortOrder(id: string, sortOrder: number): Promise<void>;

    getAllByFeatures(
        features: string[],
        environment?: string,
    ): Promise<IFeatureStrategy[]>;

    getCustomStrategiesInUseCount(): Promise<number>;
}
