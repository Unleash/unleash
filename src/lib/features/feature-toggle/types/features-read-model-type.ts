export interface IFeaturesReadModel {
    featureExists(parent: string): Promise<boolean>;
    featuresInTheSameProject(
        featureA: string,
        featureB: string,
    ): Promise<boolean>;
}
