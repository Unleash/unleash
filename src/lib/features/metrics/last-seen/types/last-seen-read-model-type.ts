export interface IFeatureLastSeenResults {
    [featureName: string]: {
        [environment: string]: {
            lastSeen: string;
        };
    };
}
export interface ILastSeenReadModel {
    getForFeature(features: string[]): Promise<IFeatureLastSeenResults>;
}
