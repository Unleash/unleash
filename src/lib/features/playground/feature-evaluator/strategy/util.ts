import * as murmurHash3 from 'murmurhash3js';

function normalizedValue(
    id: string,
    groupId: string,
    normalizer: number,
    seed = 0,
): number {
    const hash = murmurHash3.x86.hash32(`${groupId}:${id}`, seed);
    return (hash % normalizer) + 1;
}

const STRATEGY_SEED = 0;

export function normalizedStrategyValue(id: string, groupId: string): number {
    return normalizedValue(id, groupId, 100, STRATEGY_SEED);
}

const VARIANT_SEED = 86028157;

export function normalizedVariantValue(
    id: string,
    groupId: string,
    normalizer: number,
): number {
    return normalizedValue(id, groupId, normalizer, VARIANT_SEED);
}
