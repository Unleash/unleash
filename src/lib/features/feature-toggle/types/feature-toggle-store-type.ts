import {
    FeatureToggle,
    FeatureToggleDTO,
    IFeatureToggleQuery,
    IFeatureTypeCount,
    IVariant,
} from '../../../types/model';
import { FeatureToggleInsert } from '../feature-toggle-store';
import { Store } from '../../../types/stores/store';
import { LastSeenInput } from '../../metrics/last-seen/last-seen-service';
import { FeatureConfigurationClient } from './feature-toggle-strategies-store-type';
import { IFeatureProjectUserParams } from '../feature-toggle-controller';

export interface IFeatureToggleStoreQuery {
    archived: boolean;
    project: string;
    stale: boolean;
    type?: string;
}

export interface IFeatureToggleStore extends Store<FeatureToggle, string> {
    count(query?: Partial<IFeatureToggleStoreQuery>): Promise<number>;

    setLastSeen(data: LastSeenInput[]): Promise<void>;

    getProjectId(name: string): Promise<string | undefined>;

    create(project: string, data: FeatureToggleInsert): Promise<FeatureToggle>;

    update(project: string, data: FeatureToggleDTO): Promise<FeatureToggle>;

    archive(featureName: string): Promise<FeatureToggle>;

    batchArchive(featureNames: string[]): Promise<FeatureToggle[]>;

    batchStale(
        featureNames: string[],
        stale: boolean,
    ): Promise<FeatureToggle[]>;

    batchDelete(featureNames: string[]): Promise<void>;

    batchRevive(featureNames: string[]): Promise<FeatureToggle[]>;

    revive(featureName: string): Promise<FeatureToggle>;

    getAll(query?: Partial<IFeatureToggleStoreQuery>): Promise<FeatureToggle[]>;

    getAllByNames(names: string[]): Promise<FeatureToggle[]>;

    getFeatureToggleList(
        featureQuery?: IFeatureToggleQuery,
        userId?: number,
        archived?: boolean,
    ): Promise<FeatureToggle[]>;

    getArchivedFeatures(project?: string): Promise<FeatureToggle[]>;

    getPlaygroundFeatures(
        featureQuery?: IFeatureToggleQuery,
    ): Promise<FeatureConfigurationClient[]>;

    countByDate(queryModifiers: {
        archived?: boolean;
        project?: string;
        date?: string;
        range?: string[];
        dateAccessor: string;
    }): Promise<number>;

    updatePotentiallyStaleFeatures(currentTime?: string): Promise<
        {
            name: string;
            potentiallyStale: boolean;
            project: string;
        }[]
    >;

    isPotentiallyStale(featureName: string): Promise<boolean>;

    /**
     * @deprecated - Variants should be fetched from FeatureEnvironmentStore (since variants are now; since 4.18, connected to environments)
     * @param featureName
     * TODO: Remove before release 5.0
     */
    getVariants(featureName: string): Promise<IVariant[]>;

    /**
     * TODO: Remove before release 5.0
     * @deprecated - Variants should be fetched from FeatureEnvironmentStore (since variants are now; since 4.18, connected to environments)
     * @param project
     * @param featureName
     * @param newVariants
     */
    saveVariants(
        project: string,
        featureName: string,
        newVariants: IVariant[],
    ): Promise<FeatureToggle>;

    disableAllEnvironmentsForFeatures(names: string[]): Promise<void>;

    getFeatureTypeCounts(
        params: IFeatureProjectUserParams,
    ): Promise<IFeatureTypeCount[]>;

    setCreatedByUserId(batchSize: number): Promise<number | undefined>;
}
