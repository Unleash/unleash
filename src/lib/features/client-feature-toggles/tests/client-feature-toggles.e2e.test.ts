import { RoleName } from '../../../types/model';
import dbInit, { ITestDb } from '../../../../test/e2e/helpers/database-init';
import {
    IUnleashTest,
    setupAppWithCustomConfig,
} from '../../../../test/e2e/helpers/test-helper';
import getLogger from '../../../../test/fixtures/no-logger';
import { DEFAULT_ENV } from '../../../util/constants';
import { IUserWithRootRole } from '../../../types';

let app: IUnleashTest;
let db: ITestDb;
let dummyAdmin: IUserWithRootRole;

const apiClientResponse = [
    {
        name: 'test1',
        type: 'release',
        enabled: false,
        project: 'default',
        stale: false,
        strategies: [
            {
                name: 'flexibleRollout',
                constraints: [],
                parameters: { rollout: '100' },
                variants: [],
            },
        ],
        variants: [],
        description: null,
        impressionData: false,
    },
    {
        name: 'test2',
        type: 'release',
        enabled: false,
        project: 'default',
        stale: false,
        strategies: [
            {
                name: 'default',
                constraints: [
                    {
                        contextName: 'userId',
                        operator: 'IN',
                        values: ['123'],
                    },
                ],
                parameters: {},
                variants: [],
            },
        ],
        variants: [],
        description: null,
        impressionData: false,
    },
];

beforeAll(async () => {
    db = await dbInit('client_feature_toggles', getLogger);
    app = await setupAppWithCustomConfig(
        db.stores,
        {
            experimental: {
                flags: {
                    strictSchemaValidation: true,
                },
            },
        },
        db.rawDatabase,
    );

    dummyAdmin = await app.services.userService.createUser({
        name: 'Some Name',
        email: 'test@getunleash.io',
        rootRole: RoleName.ADMIN,
    });
});

afterEach(async () => {
    const all =
        await db.stores.projectStore.getEnvironmentsForProject('default');
    await Promise.all(
        all
            .filter((env) => env.environment !== DEFAULT_ENV)
            .map(async (env) =>
                db.stores.projectStore.deleteEnvironmentForProject(
                    'default',
                    env.environment,
                ),
            ),
    );
});

afterAll(async () => {
    await app.destroy();
    await db.destroy();
});

test('should fetch single feature', async () => {
    expect.assertions(1);
    await app.createFeature('test_', 'default');

    return app.request
        .get(`/api/client/features/test_`)
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((res) => {
            expect(res.body.name === 'test_').toBe(true);
        });
});

test('should support name prefix', async () => {
    expect.assertions(2);
    await app.createFeature('a_test1');
    await app.createFeature('a_test2');
    await app.createFeature('b_test1');
    await app.createFeature('b_test2');

    const namePrefix = 'b_';

    return app.request
        .get(`/api/client/features?namePrefix=${namePrefix}`)
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((res) => {
            expect(res.body.features.length).toBe(2);
            expect(res.body.features[1].name).toBe('b_test2');
        });
});

test('should support filtering on project', async () => {
    expect.assertions(2);
    await app.services.projectService.createProject(
        { name: 'projectA', id: 'projecta' },
        dummyAdmin,
    );
    await app.services.projectService.createProject(
        { name: 'projectB', id: 'projectb' },
        dummyAdmin,
    );
    await app.createFeature('ab_test1', 'projecta');
    await app.createFeature('bd_test2', 'projectb');
    return app.request
        .get(`/api/client/features?project=projecta`)
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((res) => {
            expect(res.body.features).toHaveLength(1);
            expect(res.body.features[0].name).toBe('ab_test1');
        });
});

test('should return correct data structure from /api/client/features', async () => {
    await db.rawDatabase.raw('DELETE FROM features');

    await app.createFeature('test1', 'default');
    await app.createFeature('test2', 'default');

    await app.addStrategyToFeatureEnv(
        {
            name: 'flexibleRollout',
            constraints: [],
            parameters: { rollout: '100' },
        },
        DEFAULT_ENV,
        'test1',
    );
    await app.addStrategyToFeatureEnv(
        {
            name: 'default',
            constraints: [
                { contextName: 'userId', operator: 'IN', values: ['123'] },
            ],
            parameters: {},
        },
        DEFAULT_ENV,
        'test2',
    );

    const result = await app.request
        .get('/api/client/features')
        .expect('Content-Type', /json/)
        .expect(200);

    expect(result.body.features).toEqual(apiClientResponse);
});
