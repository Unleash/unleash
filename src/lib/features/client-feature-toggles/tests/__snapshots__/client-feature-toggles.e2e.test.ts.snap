// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot from /api/client/features 1`] = `
{
  "features": [
    {
      "description": null,
      "enabled": false,
      "impressionData": false,
      "name": "test1",
      "project": "default",
      "stale": false,
      "strategies": [
        {
          "constraints": [
            {
              "caseInsensitive": false,
              "contextName": "appName",
              "inverted": false,
              "operator": "IN",
              "values": [
                "test",
              ],
            },
          ],
          "name": "flexibleRollout",
          "parameters": {
            "groupId": "test1",
            "rollout": "100",
            "stickiness": "default",
          },
          "variants": [],
        },
      ],
      "type": "release",
      "variants": [],
    },
    {
      "description": null,
      "enabled": false,
      "impressionData": false,
      "name": "test2",
      "project": "default",
      "stale": false,
      "strategies": [
        {
          "constraints": [
            {
              "contextName": "userId",
              "operator": "IN",
              "values": [
                "123",
              ],
            },
          ],
          "name": "default",
          "parameters": {},
          "variants": [],
        },
      ],
      "type": "release",
      "variants": [],
    },
  ],
  "meta": {
    "etag": ""61824cd0:11"",
    "queryHash": "61824cd0",
    "revisionId": 11,
  },
  "query": {
    "environment": "default",
    "inlineSegmentConstraints": true,
  },
  "version": 2,
}
`;
